### PR TITLE
feat(lib/runtime/wazero): implement compiler cache

### PR DESCRIPTION
**Pull request by jimjbrettj**
## Changes

Use compilation caching provided by wazero to increase sync speeds

Speedup data compared to non cache solution:

| | |
|-----| ------|
| CACHED | ⛓️ synced 7681 blocks, took: 145.12 seconds, bps: 52.93 blocks/second |
| NON CACHED | ⛓️ synced 7681 blocks, took: 749.70 seconds, bps: 10.25 blocks/second	pkg=sync |

## Tests

- run paseo/westend nodes and makes sure it works and that sync speed increases

## Issues

 #3880

## Primary Reviewer

@EclesioMeloJunior

### CODE REVIEW

**Comment by dimartiro** [comment]
Do you have any metric to check how much do we improve the sync speed with this change? I think it could be a good opportunity to revisit this issue https://github.com/ChainSafe/gossamer/issues/3823

File: dot/core/service.go [hunk 1]
```diff
@@ -312,6 +312,12 @@ func (s *Service) handleCodeSubstitution(hash common.Hash,
 		cfg.Role = 4
 	}
 
+	err = rt.CleanCache()
+	if err != nil {
+		logger.Debugf("cleaning runtime compiler cache: %v", err)
```
Comment (EclesioMeloJunior) [inline dot/core/service.go]: I think we don't need to log if we're about to return an error

File: dot/core/service.go [hunk 2]
```diff
@@ -312,6 +312,12 @@ func (s *Service) handleCodeSubstitution(hash common.Hash,
...
+		return fmt.Errorf("cleaning runtime compiler cache: %w", err)
+	}
+
```
Comment (timwu20) [inline dot/core/service.go]: Why are we cleaning the cache when handling substituted code? Shouldn't we persist this cache in the case we resume the node?

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]
Can we make the `CleanCache` a private function that belongs only to the `Instance` structure, that makes more sense given that the instance is the one who wraps the wazero instance, also call the `cleanCache` only in the `Stop` method, so we make sure that we clean the current instance cache only when the instance stop

**Comment by jimjbrettj** [comment]
> Do you have any metric to check how much do we improve the sync speed with this change? I think it could be a good opportunity to revisit this issue #3823

@dimartiro I think this is a very nice thing to have, however, I do not thing that this is something worth doing right now considering our current needs. Eclesio's work fixes memory issues and gives us correctness and then this is a speedup on that. I think for now I have more pressing things to work on, but in the future this would be nice to have. I will however still add some basic metrics I collect to add a bit more context into the speed up

File: dot/core/service.go (hunk 2 above)
Comment (jimjbrettj) [reply dot/core/service.go]: We need to close the cache to prevent a memory leak, however I do not think we need to remove the dir so I will make this change

Comment (jimjbrettj) [reply dot/core/service.go]: And you are right than in this case we do not need to clean the cache. This PR has been modified to only close the cache when the instance is pruned

File: dot/core/service.go (hunk 1 above)
Comment (jimjbrettj) [reply dot/core/service.go]: this has been removed

File: lib/runtime/wazero/instance.go [hunk 3]
```diff
@@ -412,8 +422,19 @@ func NewInstance(code []byte, cfg Config) (instance *Instance, err error) {
 	logger.Info("instantiating a runtime!")
 	logger.Patch(log.SetLevel(cfg.LogLvl), log.SetCallerFunc(true))
 
+	// Prepare a cache directory.
+	cacheDir, err := os.MkdirTemp("", "wazeroCache")
```
Comment (EclesioMeloJunior) [inline lib/runtime/wazero/instance.go]: saves the `cacheDir` to delete it once we close the instance, looks like it does not remove the dir once it closes the cache https://github.com/tetratelabs/wazero/blob/253c0346e4c8d7bd604304447233c1a3ce0a71f7/cache.go#L70

Comment (jimjbrettj) [reply lib/runtime/wazero/instance.go]: yeah exactly, I talked with @timwu20 and we were thinking we can just keep it for if we restart the node

File: lib/runtime/wazero/instance.go [hunk 4]
```diff
@@ -36,13 +37,19 @@ const runtimeContextKey = contextKey("runtime.Context")
...
+	config wazero.RuntimeConfig
+	cache  wazero.CompilationCache
+}
+
 // Instance backed by wazero.Runtime
 type Instance struct {
 	Runtime  wazero.Runtime
 	Module   api.Module
 	Context  *runtime.Context
+	code     []byte
```
Comment (axaysagathiya) [inline lib/runtime/wazero/instance.go]: Can you use a more meaningful field name here? 

If I am correct, it's a wasm bytecode. Maybe you can rename it to `wasmByteCode.`

File: lib/runtime/wazero/instance.go [hunk 5]
```diff
@@ -100,15 +107,13 @@ func NewInstanceFromTrie(t trie.Trie, cfg Config) (*Instance, error) {
 	return NewInstance(code, cfg)
 }
 
-// NewInstance instantiates a runtime from raw wasm bytecode
```
Comment (axaysagathiya) [inline lib/runtime/wazero/instance.go:103]: You could have kept this comment. It will help to know what the `code` argument is.

File: lib/runtime/wazero/instance.go (hunk 3 above)
Comment (timwu20) [reply lib/runtime/wazero/instance.go]: We should be putting this in `~/.gossamer` then if we plan to keep it.

File: lib/runtime/wazero/instance.go [hunk 6]
```diff
@@ -396,35 +401,47 @@ func NewInstance(code []byte, cfg Config) (instance *Instance, err error) {
...
-		return nil, fmt.Errorf("wazero error: nil global for __heap_base")
-	}
+	return mod, rt, nil
+}
 
-	heapBase := api.DecodeU32(encodedHeapBase.Get())
-	// hb = runtime.DefaultHeapBase
+// NewInstance instantiates a runtime from raw wasm bytecode
+func NewInstance(code []byte, cfg Config) (instance *Instance, err error) {
+	logger.Info("instantiating a runtime!")
```
Comment (timwu20) [inline lib/runtime/wazero/instance.go]: maybe make this debug? Seems pretty obvious we are instantiating a runtime.

File: lib/runtime/wazero/instance.go [hunk 7]
```diff
@@ -899,4 +937,9 @@ func (in *Instance) Stop() {
 	if err != nil {
 		log.Errorf("runtime failed to close: %v", err)
 	}
+
+	err = in.metadata.cache.Close(context.Background())
+	if err != nil {
+		log.Errorf("closing the wazero compilation cache: %v", err)
```
Comment (timwu20) [inline lib/runtime/wazero/instance.go:943]: why not return this error?

File: lib/runtime/wazero/instance_test.go [hunk 8]
```diff
@@ -1096,7 +1095,7 @@ func newTrieFromPairs(t *testing.T, filename string) trie.Trie {
 		entries[pairArr[0].(string)] = pairArr[1].(string)
 	}
 
-	tr, err := inmemory.LoadFromMap(entries, trie.V0)
```
Comment (timwu20) [inline lib/runtime/wazero/instance_test.go:1099]: unnecessary diff

**Review by timwu20 — CHANGES_REQUESTED** [review]

File: lib/runtime/wazero/instance.go (hunk 3 above)
Comment (jimjbrettj) [reply lib/runtime/wazero/instance.go]: making this change, however, we have moved to using `xdg.DataHome` rather than `~/.gossamer` so using that instead

File: lib/runtime/wazero/instance.go (hunk 5 above)
Comment (jimjbrettj) [reply lib/runtime/wazero/instance.go:103]: This comment is still there, just in a new place

File: lib/runtime/wazero/instance.go (hunk 7 above)
Comment (jimjbrettj) [reply lib/runtime/wazero/instance.go:943]: The current stop function signature did not return an error, it just logged if there was an error closing so I just kept with what was being done before. I can update this if desired though

File: lib/runtime/wazero/instance_test.go (hunk 8 above)
Comment (jimjbrettj) [reply lib/runtime/wazero/instance_test.go:1099]: I see this comment on the previous PR (lost when we combined them) https://github.com/ChainSafe/gossamer/pull/3865#discussion_r1552423920

So maybe there is context to this change I am missing. @dimartiro could you add some context here? else I will delete this diff

Comment (dimartiro) [reply lib/runtime/wazero/instance_test.go:1099]: This is just to remove the old import and use the new one with the alias

Comment (jimjbrettj) [reply lib/runtime/wazero/instance_test.go:1099]: okay, i will keep then

**Comment by jimjbrettj** [comment]
I did not see any noticeable improvement with the cache dir, and it was breaking all CI, and with @ec2 changes implementing more caching I do no think it is needed so I removed it

**Review by EclesioMeloJunior — APPROVED** [review]

File: chain/paseo/chain-spec-raw.json [hunk 9]
```diff
@@ -6,7 +6,19 @@
     "/dns/paseo.bootnode.amforc.com/tcp/30333/wss/p2p/12D3KooWFD81HC9memUwuGMLvhDDEfmXjn6jC4n7zyNs3vToXapS",
     "/dns/paseo.bootnode.amforc.com/tcp/30344/p2p/12D3KooWFD81HC9memUwuGMLvhDDEfmXjn6jC4n7zyNs3vToXapS",
     "/dns/boot.stake.plus/tcp/43334/wss/p2p/12D3KooWNhgAC3hjZHxaT52EpPFZohkCL1AHFAijqcN8xB9Rwud2",
-    "/dns/boot.stake.plus/tcp/43333/p2p/12D3KooWNhgAC3hjZHxaT52EpPFZohkCL1AHFAijqcN8xB9Rwud2"
+    "/dns/boot.stake.plus/tcp/43333/p2p/12D3KooWNhgAC3hjZHxaT52EpPFZohkCL1AHFAijqcN8xB9Rwud2",
```
Comment (dimartiro) [inline chain/paseo/chain-spec-raw.json:9]: Are we adding these nodes because we are testing this change in paseo? I think this could be in a different PR

**Review by dimartiro — APPROVED** [review]
lgtm

File: chain/paseo/chain-spec-raw.json (hunk 9 above)
Comment (jimjbrettj) [reply chain/paseo/chain-spec-raw.json:9]: Well yes, but often the node will not sync without this change. So I think okay to keep here since small and allows testing

File: lib/runtime/wazero/instance.go [hunk 10]
```diff
@@ -100,15 +108,13 @@ func NewInstanceFromTrie(t trie.Trie, cfg Config) (*Instance, error) {
...
 }
 
-// NewInstance instantiates a runtime from raw wasm bytecode
-func NewInstance(code []byte, cfg Config) (instance *Instance, err error) {
-	logger.Info("instantiating a runtime!")
-	logger.Patch(log.SetLevel(cfg.LogLvl), log.SetCallerFunc(true))
-
-	ctx := context.Background()
-	rt := wazero.NewRuntime(ctx)
+func newRuntimeInstance(ctx context.Context,
```
Comment (timwu20) [inline lib/runtime/wazero/instance.go]: I feel like this function name should be changed given it doesn't return an `Instance` anymore.

File: lib/runtime/wazero/instance.go [hunk 11]
```diff
@@ -36,13 +36,21 @@ const runtimeContextKey = contextKey("runtime.Context")
 
 var _ runtime.Instance = &Instance{}
 
+type cacheMetadata struct {
```
Comment (timwu20) [inline lib/runtime/wazero/instance.go]: maybe call this `wazeroMeta`? It's not just the cache anymore.

File: lib/runtime/wazero/instance.go [hunk 12]
```diff
@@ -36,13 +36,21 @@ const runtimeContextKey = contextKey("runtime.Context")
...
+	config      wazero.RuntimeConfig
+	cache       wazero.CompilationCache
+	hostModule  wazero.CompiledModule
```
Comment (timwu20) [inline lib/runtime/wazero/instance.go]: why do we need to store this?

Comment (jimjbrettj) [reply lib/runtime/wazero/instance.go]: This is a nice find, i dont think we need to

File: lib/runtime/wazero/instance.go (hunk 10 above)
Comment (jimjbrettj) [reply lib/runtime/wazero/instance.go]: updated: renamed to `newRuntime` since a runtime and associated data are still being created. Let me know if you prefer something else

Comment (EclesioMeloJunior) [reply lib/runtime/wazero/instance.go]: Agreed, maybe change to compiled modules

Comment (jimjbrettj) [reply lib/runtime/wazero/instance.go]: Hm but it returns one runtime and one module (since we don't cache host module anymore). Maybe something like `newRuntimeWithGuestModule`?

**Review by timwu20 — APPROVED** [review]

File: lib/runtime/wazero/instance.go (hunk 12 above)
Comment (EclesioMeloJunior) [reply lib/runtime/wazero/instance.go]: yeah, we should only keep the guest compiled module since this is the piece that are going to be instantiated on every exec.
